File: DSA/algorithms/grid_traveler.py
# recursion
def grid_traveler(m, n):
    if (m==1) and (n==1):
        return 1
    if (m==0) or (n==0):
        return 0

    return grid_traveler(m-1, n) + grid_traveler(m ,n-1)

# memoization
def grid_traveler_mem(m, n, dict={}):
    key = str(m) + ',' + str(n)
    if key in dict.keys():
        return dict[key]
    if (m==1) and (n==1):
        return 1
    if (m==0) or (n==0):
        return 0

    dict[key] = grid_traveler_mem(m-1, n, dict) + grid_traveler_mem(m ,n-1, dict)
    return dict[key]
    
# tabulation
def grid_traveler_tab(m, n):
    dp = [[0 for i in range(n+1)] for j in range(m+1)]
    dp[1][1] = 1

    for i in range(m+1):
        for j in range(n+1):
            current = dp[i][j]
            if (i+1) <= m:
                dp[i+1][j] += current
            if (j+1) <= n:
                dp[i][j+1] += current

    return dp[m][n]

if __name__ == '__main__':
    print(grid_traveler(1,1))
    print(grid_traveler(2,3))
    print(grid_traveler(3,2))
    print(grid_traveler(3,3))

    print(grid_traveler_mem(1,1))
    print(grid_traveler_mem(2,3))
    print(grid_traveler_mem(3,2))
    print(grid_traveler_mem(3,3))
    print(grid_traveler_mem(19,19))
    
    print(grid_traveler_tab(1,1))
    print(grid_traveler_tab(2,3))
    print(grid_traveler_tab(3,2))
    print(grid_traveler_tab(3,3))
    print(grid_traveler_tab(19,19))

File: DSA/algorithms/gridTraveler.js
// recursion
const gridTraveler = (m, n) => {
	if (m===1 && n===1) return 1
	if (m===0 || n===0) return 0
	return gridTraveler(m-1, n) + gridTraveler(m, n-1)
}

// memoization
function gridTravelerMem(m, n, memo={}) {
	const key = m + ',' + n

	if (key in memo) return memo[key]
	if (m===1 && n===1) return 1
	if (m===0 || n===0) return 0

	memo[key] = gridTravelerMem(m-1, n, memo) + gridTravelerMem(m, n-1, memo)
	return memo[key]
}

// tabulation
function gridTravelerTab(m, n) {
	let dp = Array(m+1).fill().map(() => Array(n+1).fill(0))
	dp[1][1] = 1

	for (let i=0; i<=m; i++) {
		for (let j=0; j<=n; j++) {
			let current = dp[i][j]
			if (j+1<=n) dp[i][j+1] += current
			if (i+1<=m) dp[i+1][j] += current
		}
	}

	return dp[m][n]
}

console.log(gridTraveler(1,1))
console.log(gridTraveler(2,3))
console.log(gridTraveler(3,2))
console.log(gridTraveler(3,3))

console.log(gridTravelerMem(1,1))
console.log(gridTravelerMem(2,3))
console.log(gridTravelerMem(3,2))
console.log(gridTravelerMem(3,3))
console.log(gridTravelerMem(19,19))

console.log(gridTravelerTab(1,1))
console.log(gridTravelerTab(2,3))
console.log(gridTravelerTab(3,2))
console.log(gridTravelerTab(3,3))
console.log(gridTravelerTab(19,19))
